File: packages/mobile/ios/VideoProvider.swift
//
//  VideoProvider.swift
//  HomebasePhotos
//
//  Created by Stef Coenen on 11/09/2024.
//

import Foundation
import AVFoundation // for video handling
import ImageIO
import MobileCoreServices
import ffmpegkit

class VideoProvider {
  static let defaultPayloadKey = "dflt_key"
  static let encryptMedia = true
  static let photoDrive = TargetDrive(alias: "6483b7b1f71bd43eb6896c86148668cc", type: "2af68fe72fb84896f39f97c59d60813a")
  static let ownerOnlyACL = AccessControlList(requiredSecurityGroup: .owner)
  static let tinyThumbInstruction = ImageResizer.ResizeInstruction(width: 20, height: 20, quality: 10, format: "jpeg")
  static let defaultImageSizes = [
    ImageResizer.ResizeInstruction(width: 300, height: 300, quality: 95, format: "jpeg"),
    ImageResizer.ResizeInstruction(width: 1200, height: 1200, quality: 95, format: "jpeg")
  ]

  static func uploadMedia(dotYouClient: DotYouClient, filePath: String, timestampInMs: Int64, mimeType: String, identifier: String?, width: Int, height: Int, forceLowerQuality: Bool) async throws -> UploadResult {
    let instructions = UploadInstructionSet(storageOptions: StorageOptions(drive: photoDrive), transitOptions: nil, transferIv: nil, manifest: nil)

    let fileName = (filePath as NSString).lastPathComponent
    let uniqueId = try toGuidId(input: identifier ?? "\(fileName)_\(width)x\(height)")

    // Grab thumbnail
    let videoThumbnail = try await grabVideoThumbnail(filePath: filePath)


    let tinyThumb = try ImageResizer.resizeImage(imageSource: videoThumbnail!, instruction: tinyThumbInstruction, key: defaultPayloadKey, keepDimensions: true)
    var previewThumbnail: EmbeddedThumb?
    if(tinyThumb != nil) {
      previewThumbnail = EmbeddedThumb(pixelHeight: tinyThumb!.pixelHeight, pixelWidth: tinyThumb!.pixelWidth, contentType: tinyThumb!.contentType, content: CryptoUtil.base64Encode(stream: tinyThumb!.inputStream.stream)!)
    }

    let thumbnails = try ImageResizer.resizeImage(imageSource: videoThumbnail!, instructions: defaultImageSizes, key: defaultPayloadKey)

    let metadata = UploadFileMetadata<String>(
      allowDistribution: false,
      isEncrypted: encryptMedia,
      accessControlList: ownerOnlyACL,
      appData: UploadAppFileMetaData(uniqueId: uniqueId, tags: [], fileType: 0, dataType: 0, userDate: timestampInMs, groupId: nil, archivalStatus: 0, content: "{\"originalFileName\":\"" + fileName + "\"}", previewThumbnail: previewThumbnail),
      referencedFile: nil
    )

    // Compress and segment video
    let keyHeader = encryptMedia ? DriveFileUploadProvider.generateKeyHeader() : nil
    let videoData = try await compressAndSegmentVideo(filePath: filePath, compress: forceLowerQuality, keyHeader: keyHeader)
    let payloads: [PayloadBase]
    if let segments = videoData.segments {
      let playlistContent = try String(contentsOfFile: videoData.video!.filepath, encoding: .utf8)

      // Handle HLS segments
      let metadata: [String: Any] = [
              "isSegmented": true,
              "mimeType": "application/vnd.apple.mpegurl",
              "hlsPlaylist": playlistContent
      ];

      // Convert dictionary to JSON data
      let jsonMetadata = try JSONSerialization.data(withJSONObject: metadata, options: .prettyPrinted)

      let payloadStream = PayloadFile(descriptorContent: String(data: jsonMetadata, encoding: .utf8), previewThumbnail: nil, contentType: "video/mp2t", key: defaultPayloadKey, filePath: segments.filepath, skipEncryption: true, iv: keyHeader?.iv)
      payloads = [payloadStream]
    } else {
      // Single video file
      let metadata: String = "{ mimeType: 'video/mp4', isSegmented: false, }";

      let payloadFile = PayloadFile(descriptorContent: metadata, previewThumbnail: nil, contentType: mimeType, key: defaultPayloadKey, filePath: videoData.video!.filepath, skipEncryption: false)
      payloads = [payloadFile]
    }

    return try await DriveFileUploadProvider.uploadFile(dotYouClient: dotYouClient, instructions: instructions, metadata: metadata, payloads: payloads, thumbnails: thumbnails, aesKey: keyHeader?.aesKey)
  }

  static func toGuidId(input: String) throws -> String {
    var digest = [UInt8](repeating: 0, count: Int(CC_MD5_DIGEST_LENGTH))
    if let data = input.data(using: .utf8) {
      _ = data.withUnsafeBytes {
        CC_MD5($0.baseAddress, CC_LONG(data.count), &digest)
      }
    }

    let uuid = UUID(digest: digest)
    return uuid.uuidString
  }

  static func grabVideoThumbnail(filePath: String) async throws -> CGImageSource? {
    // Use AVFoundation to extract a thumbnail from the video
    let asset = AVAsset(url: URL(fileURLWithPath: filePath))
    let imageGenerator = AVAssetImageGenerator(asset: asset)
    imageGenerator.appliesPreferredTrackTransform = true // Ensure correct orientation
    let time = CMTime(seconds: 1, preferredTimescale: 60) // Grab a frame from 1 second in

    do {
      let cgImage = try imageGenerator.copyCGImage(at: time, actualTime: nil)

      // Convert CGImage to Data
      let imageData = NSMutableData()
      if let destination = CGImageDestinationCreateWithData(imageData as CFMutableData, kUTTypeJPEG, 1, nil) {
        CGImageDestinationAddImage(destination, cgImage, nil)
        CGImageDestinationFinalize(destination)
      } else {
        throw NSError(domain: "com.yourdomain.app", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to create image destination"])
      }

      // Create CGImageSource from Data
      if let imageSource = CGImageSourceCreateWithData(imageData as CFData, nil) {
        return imageSource
      } else {
        throw NSError(domain: "com.yourdomain.app", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to create CGImageSource from image data"])
      }
    } catch {
      throw error
    }
  }

  static func compressAndSegmentVideo(filePath: String, compress: Bool, keyHeader: KeyHeader?) async throws -> VideoData {
    let outputDir = FileManager.default.temporaryDirectory
    let inputVideoUrl = URL(fileURLWithPath: filePath)

    let compressedVideoUrl = compress ? try await compressVideo(inputUrl: inputVideoUrl, outputUrl: outputDir.appendingPathComponent("compressed.mp4")) : inputVideoUrl;
    let (playlistUrl, segmentsUrl) = try await segmentVideoToHLS(inputUrl: compressedVideoUrl, outputDirectory: outputDir, keyHeader: keyHeader)

    print("playlist: \(playlistUrl.path) & segments \(segmentsUrl.path)")
    return VideoData(video: VideoFile(filepath: playlistUrl.path), segments: VideoSegments(filepath: segmentsUrl.path))
  }


  static func compressVideo(inputUrl: URL, outputUrl: URL) async throws -> URL {
    let asset = AVAsset(url: inputUrl)

    guard let exportSession = AVAssetExportSession(asset: asset, presetName: AVAssetExportPreset1280x720) else {
      throw NSError(domain: "com.yourdomain.app", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to create export session"])
    }

    exportSession.outputFileType = .mp4
    exportSession.outputURL = outputUrl
    exportSession.shouldOptimizeForNetworkUse = true

    return try await withCheckedThrowingContinuation { continuation in
      exportSession.exportAsynchronously {
        switch exportSession.status {
        case .completed:
          continuation.resume(returning: outputUrl)
        case .failed, .cancelled:
          continuation.resume(throwing: exportSession.error ?? NSError(domain: "com.yourdomain.app", code: -1, userInfo: [NSLocalizedDescriptionKey: "Video compression failed"]))
        default:
          break
        }
      }
    }
  }

  static func segmentVideoToHLS(inputUrl: URL, outputDirectory: URL, keyHeader: KeyHeader?) async throws -> (playlistUrl: URL, segmentsUrl: URL) {
    let id = UUID().uuidString
    let playlistFileName = "output-\(id).m3u8"
    let segmentsFileName = "output-\(id).ts"
    let playlistUrl = outputDirectory.appendingPathComponent(playlistFileName)
    let segmentsUrl = outputDirectory.appendingPathComponent(segmentsFileName)

    var encryptionCommand = ""

    // If encryption is enabled, create a key file and use it in the ffmpeg command
    if let keyHeader = keyHeader {
      let keyInfoFile = try await createKeyInfoFile(directory: outputDirectory, keyHeader: keyHeader)
      encryptionCommand = "-hls_key_info_file \(keyInfoFile.path)"
    }

    // ffmpeg command to segment the video into HLS
    let command = "-i \(inputUrl.path) -codec copy \(encryptionCommand) -hls_time 6 -hls_list_size 0 -f hls -hls_flags single_file \(playlistUrl.path)"
    print("ffmpeg command: \(command)")

    return try await withCheckedThrowingContinuation { continuation in
      FFmpegKit.executeAsync(command) { session in
        let state = session!.getState()
        let returnCode = session!.getReturnCode()

        if state == .failed || !(returnCode?.isValueSuccess())! {
          let error = NSError(domain: "com.yourdomain.app", code: -1, userInfo: [NSLocalizedDescriptionKey: "ffmpeg process failed with state: \(state) and returnCode: \(String(describing: returnCode))"])
          continuation.resume(throwing: error)
        } else {
          continuation.resume(returning: (playlistUrl, segmentsUrl))
        }
      }
    }
  }

  // Helper function to create a key info file for encryption (if required)
  static func createKeyInfoFile(directory: URL, keyHeader: KeyHeader) async throws -> URL {
    let id = UUID().uuidString
    let keyFileUrl = directory.appendingPathComponent("hls-encryption\(id).key")
    let keyInfoFileUrl = directory.appendingPathComponent("hls-key-info\(id).txt")

    // Convert aesKey and iv (Data) to base64 or hex strings
    let ivHexString = keyHeader.iv.map { String(format: "%02hhx", $0) }.joined()  // Convert to hex string

    // Write aesKey to the key file in base64 format
    try keyHeader.aesKey.write(to: keyFileUrl)

    // Create the key info content
    let keyInfo = "http://example.com/path/to/encryption.key\n\(keyFileUrl.path)\n\(ivHexString)"

    // Write keyInfo to the key-info file
    try keyInfo.write(to: keyInfoFileUrl, atomically: true, encoding: .utf8)

    return keyInfoFileUrl
  }

  // Helper Structures
  struct VideoData {
    var video: VideoFile?
    var segments: VideoSegments?
  }

  struct VideoFile {
    var filepath: String
  }

  struct VideoSegments {
    var filepath: String
  }
}

File: packages/mobile/android/app/src/main/java/id/homebase/photos/mediasync/VideoProvider.java
package id.homebase.photos.mediasync;

import static id.homebase.lib.core.file.DriveFileUploadProvider.uploadFile;

import android.content.Context;
import android.util.Log;

import java.io.File;
import java.io.IOException;
import java.nio.file.Paths;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

import id.homebase.lib.core.DotYouClient;
import id.homebase.lib.core.crypto.CryptoUtil;
import id.homebase.lib.core.file.types.AccessControlList;
import id.homebase.lib.core.file.types.ArchivalStatus;
import id.homebase.lib.core.file.types.EmbeddedThumb;
import id.homebase.lib.core.file.types.KeyHeader;
import id.homebase.lib.core.file.types.KeyHeaderGenerator;
import id.homebase.lib.core.file.types.SecurityGroupType;
import id.homebase.lib.core.file.types.StorageOptions;
import id.homebase.lib.core.file.types.TargetDrive;
import id.homebase.lib.core.file.types.UploadAppFileMetaData;
import id.homebase.lib.core.file.types.UploadFileMetadata;
import id.homebase.lib.core.file.types.UploadInstructionSet;
import id.homebase.lib.core.file.types.UploadResult;
import id.homebase.lib.core.file.types.payloadorthumbnailbase.PayloadBase;
import id.homebase.lib.core.file.types.payloadorthumbnailbase.PayloadFile;
import id.homebase.lib.core.file.types.payloadorthumbnailbase.ThumbnailBase;
import id.homebase.lib.core.file.types.payloadorthumbnailbase.ThumbnailStream;
import id.homebase.photos.mediasync.types.VideoData;
import id.homebase.photos.mediasync.types.VideoFile;
import id.homebase.photos.mediasync.types.VideoSegments;
import com.arthenica.ffmpegkit.FFmpegKit;
import com.arthenica.ffmpegkit.FFprobeKit;
import com.arthenica.ffmpegkit.ReturnCode;

import org.json.JSONObject;

public class VideoProvider {
    private static final String DEFAULT_PAYLOAD_KEY = "dflt_key";
    private static final boolean ENCRYPT_MEDIA = true;
    private static final TargetDrive PHOTO_DRIVE = new TargetDrive("6483b7b1f71bd43eb6896c86148668cc", "2af68fe72fb84896f39f97c59d60813a");
    private static final AccessControlList OWNER_ONLY_ACL = new AccessControlList(SecurityGroupType.OWNER);
    private static final ImageResizer.ResizeInstruction TINY_THUMB_INSTRUCTION = new ImageResizer.ResizeInstruction(20, 20, 10, "jpeg");
    private static final ImageResizer.ResizeInstruction[] DEFAULT_IMAGE_SIZES = new ImageResizer.ResizeInstruction[]{
            new ImageResizer.ResizeInstruction(300, 300, 95, "jpeg"),
            new ImageResizer.ResizeInstruction(1200, 1200, 95, "jpeg"),
    };

    private final Context context;

    public VideoProvider(Context context) {
        this.context = context;
    }

    public UploadResult uploadMedia(DotYouClient dotYouClient, String filePath, Long timestampInMs, String mimeType, String identifier, String width, String height, boolean forceLowerQuality) throws Exception {
        UploadInstructionSet instructions = new UploadInstructionSet(new StorageOptions(PHOTO_DRIVE));

        // Retrieve the latest photo information
        String fileName = Paths.get(filePath).getFileName().toString();
        String uniqueId = toGuidId(identifier != null ? identifier : fileName + "_" + width + "x" + height);

        // Grab thumbnails for the video
        String videoThumbnailPath = grabVideoThumbnail(filePath).get();

        // Generate thumbnails
        ThumbnailStream tinyThumb = ImageResizer.resizeImage(videoThumbnailPath, TINY_THUMB_INSTRUCTION, DEFAULT_PAYLOAD_KEY, true);
        EmbeddedThumb previewThumbnail = new EmbeddedThumb(tinyThumb.getPixelHeight(), tinyThumb.getPixelWidth(), TINY_THUMB_INSTRUCTION.format, tinyThumb.getBase64());
        List<ThumbnailBase> thumbnails = new ArrayList<>(ImageResizer.resizeImage(videoThumbnailPath, List.of(DEFAULT_IMAGE_SIZES), DEFAULT_PAYLOAD_KEY));

        UploadFileMetadata<String> metadata = new UploadFileMetadata<>(false, ENCRYPT_MEDIA, OWNER_ONLY_ACL, new UploadAppFileMetaData<>(uniqueId, new String[0], 0, 0, timestampInMs, null, ArchivalStatus.None, "{\"originalFileName\":\"" + fileName + "\"}", previewThumbnail), null, null);

        // Compress and segment video
        // Handle compression and segmentation
        KeyHeader keyHeader = ENCRYPT_MEDIA ? KeyHeaderGenerator.generateKeyHeader() : null;
        VideoData videoData = compressAndSegmentVideo(filePath, forceLowerQuality, keyHeader);

        List<PayloadBase> payloads;
            if (videoData.getSegments() != null) {
                String playlistContent = new String(java.nio.file.Files.readAllBytes(Paths.get(videoData.getVideo().getFilePath())));

                // Handle HLS segments
                JSONObject metadataJson = new JSONObject();
                metadataJson.put("isSegmented", true);
                metadataJson.put("mimeType", "application/vnd.apple.mpegurl");
                metadataJson.put("hlsPlaylist", playlistContent);

                // Convert JSONObject to String
                String metadataJsonString = metadataJson.toString();

            PayloadFile payloadStream = new PayloadFile(DEFAULT_PAYLOAD_KEY, videoData.getSegments().getFilePath(), previewThumbnail, "video/mp2t",  metadataJsonString, true, keyHeader != null ? keyHeader.iv() : null);
            payloads = List.of(payloadStream);
        } else {
            // Handle single video file
            String metadataStr = "{ \"mimeType\": \"video/mp4\", \"isSegmented\": false }";

            PayloadFile payloadFile = new PayloadFile(DEFAULT_PAYLOAD_KEY, videoData.getVideo().getFilePath(), previewThumbnail, "video/mp4", metadataStr);
            payloads = List.of(payloadFile);
        }

        if(keyHeader != null) {
            return uploadFile(dotYouClient, instructions, metadata, payloads, thumbnails, keyHeader.aesKey());
        } else {
            return uploadFile(dotYouClient, instructions, metadata, payloads, thumbnails, ENCRYPT_MEDIA);
        }
    }

    // Function to generate video thumbnail using FFmpeg
    public CompletableFuture<String> grabVideoThumbnail(String filePath) {
        File outputDir = context.getCacheDir();
        String fileName = "thumbnail-" + java.util.UUID.randomUUID();

        String argThumbnailPath = new File(outputDir, fileName + "%04d.png").getAbsolutePath();
        String thumbnailPath = new File(outputDir, fileName + "0001.png").getAbsolutePath();

        return CompletableFuture.supplyAsync(() -> {
            String command = String.format("-i %s -frames:v 1 %s", filePath, argThumbnailPath);
            try {
                var session = FFmpegKit.execute(command);
                ReturnCode returnCode = session.getReturnCode();
                if (returnCode.isValueSuccess()) {
                    return thumbnailPath;  // Successfully generated the thumbnail
                } else {
                    throw new RuntimeException("Failed to generate thumbnail: " + session.getAllLogsAsString());
                }
            } catch (Exception e) {
                throw new RuntimeException("Error generating thumbnail", e);
            }
        });
    }

    // Function to compress and segment video
    public VideoData compressAndSegmentVideo(String filePath, boolean compress, KeyHeader keyHeader) throws Exception {
        File outputDir = context.getCacheDir();
        File inputVideoFile = new File(filePath);

        File compressedVideoFile = compress ? compressVideo(inputVideoFile, new File(outputDir, "compressed.mp4")).get() : inputVideoFile;
        File[] hlsFiles = segmentVideoToHLS(compressedVideoFile, outputDir, keyHeader).get();

        return new VideoData(new VideoFile(hlsFiles[0].getAbsolutePath()), new VideoSegments(hlsFiles[1].getAbsolutePath()));
    }

    // Function to compress video using FFmpeg
    public CompletableFuture<File> compressVideo(File inputFile, File outputFile) {
        return CompletableFuture.supplyAsync(() -> {
            String command = String.format("-i %s -preset fast -crf 23 %s", inputFile.getAbsolutePath(), outputFile.getAbsolutePath());
            try {
                var session = FFmpegKit.execute(command);
                ReturnCode returnCode = session.getReturnCode();
                if (returnCode.isValueSuccess()) {
                    return outputFile;  // Successfully compressed the video
                } else {
                    throw new RuntimeException("Failed to compress video: " + session.getAllLogsAsString());
                }
            } catch (Exception e) {
                throw new RuntimeException("Error compressing video", e);
            }
        });
    }

    // Function to get the codec of a video file
    public CompletableFuture<String> getVideoCodec(String inputFilePath) {
        CompletableFuture<String> codecFuture = new CompletableFuture<>();

        // Command to retrieve the codec of the video file using ffprobe
        String command = "-v error -select_streams v:0 -show_entries stream=codec_name -of default=noprint_wrappers=1:nokey=1 " + inputFilePath;

        // Execute FFmpeg command
        FFprobeKit.executeAsync(command, session -> {
            ReturnCode returnCode = session.getReturnCode();
            if (ReturnCode.isSuccess(returnCode)) {
                // Success: Retrieve codec from output
                String output = session.getOutput();
                codecFuture.complete(output.trim());
            } else {
                // Failure: Handle error
                String errorMessage = session.getFailStackTrace();
                codecFuture.completeExceptionally(new Exception("Failed to get codec: " + errorMessage));
            }
        });

        return codecFuture;
    }

    // Function to segment video to HLS using FFmpeg
    public CompletableFuture<File[]> segmentVideoToHLS(File inputFile, File outputDir, KeyHeader keyHeader) {
        return CompletableFuture.supplyAsync(() -> {
            String randomId = UUID.randomUUID().toString();
            String playlistFileName = "output-" + randomId + ".m3u8";
            String segmentsFileName = "output-" + randomId + ".ts";
            File playlistFile = new File(outputDir, playlistFileName);
            File segmentsFile = new File(outputDir, segmentsFileName);

            String encryptionCommand = "";
            try {
                if (keyHeader != null) {
                    File keyInfoFile = createKeyInfoFile(outputDir, keyHeader);  // This method will need to be implemented
                    encryptionCommand = "-hls_key_info_file " + keyInfoFile.getAbsolutePath();
                }

                 // Extract video rotation metadata
                String rotation = extractRotationMetadata(inputFile); // Implement this method to extract rotation metadata

                // Define the transpose filter based on the rotation metadata
                String rotationCommand = "";
                if (rotation != null) {
                    switch (rotation) {
                        case "90":
                            rotationCommand = "-vf \"transpose=1\""; // Rotate 90 degrees clockwise
                            break;
                        case "180":
                            rotationCommand = "-vf \"transpose=2,transpose=2\""; // Rotate 180 degrees
                            break;
                        case "270":
                            rotationCommand = "-vf \"transpose=2\""; // Rotate 270 degrees clockwise (or 90 degrees counter-clockwise)
                            break;
                        default:
                            break; // No rotation needed for 0 or unknown rotation values
                    }
                }

                String command;
                if(getVideoCodec(inputFile.getAbsolutePath()).get().equals("h264")) {
                    // Command to segment the video to HLS using FFmpeg
                    command = String.format("-i %s -codec copy %s %s -hls_time 6 -hls_list_size 0 -f hls -hls_flags single_file %s",
                            inputFile.getAbsolutePath(), encryptionCommand, rotationCommand, playlistFile.getAbsolutePath());
                } else {
                    // Command to segment the video to HLS using FFmpeg
                    command = String.format("-i %s -c:v libx264 -preset fast -crf 23 -c:a aac %s %s -hls_time 6 -hls_list_size 0 -f hls -hls_flags single_file %s",
                            inputFile.getAbsolutePath(), encryptionCommand, rotationCommand, playlistFile.getAbsolutePath());
                }

                // Execute the FFmpeg command using FFmpegKit
                var session = FFmpegKit.execute(command);
                ReturnCode returnCode = session.getReturnCode();

                if (returnCode.isValueSuccess()) {
                    // Successfully segmented the video
                    return new File[]{playlistFile, segmentsFile};
                } else {
                    throw new RuntimeException("Failed to segment video to HLS: " + session.getAllLogsAsString());
                }
            } catch (Exception e) {
                throw new RuntimeException("Error during HLS segmentation", e);
            }
        });
    }

    // Example method to extract rotation metadata
    private String extractRotationMetadata(File inputFile) throws IOException {
        // Use FFmpeg or MediaMetadataRetriever to extract the rotation metadata
        String rotation = null;
        String command = String.format("-i %s -select_streams v:0 -show_entries stream_tags=rotate -of csv=p=0", inputFile.getAbsolutePath());

        // Execute FFmpeg command to get the rotation metadata
        var session = FFmpegKit.execute(command);
        if (session.getReturnCode().isValueSuccess()) {
            String output = session.getAllLogsAsString().trim();
            if (!output.isEmpty()) {
                rotation = output;
            }
        }
        return rotation;
    }

    private String bytesToHex(byte[] bytes) {
        StringBuilder hexString = new StringBuilder();
        for (byte b : bytes) {
            String hex = String.format("%02x", b);  // Converts byte to hex, padding with a leading zero if necessary
            hexString.append(hex);
        }
        return hexString.toString();
    }

    // Helper function to create key info file for encryption
    public File createKeyInfoFile(File directory, KeyHeader keyHeader) throws IOException {
        String id = java.util.UUID.randomUUID().toString();
        File keyFile = new File(directory, "hls-encryption-" + id + ".key");
        File keyInfoFile = new File(directory, "hls-key-info-" + id + ".txt");

        // Write AES key to the key file in raw format
        java.nio.file.Files.write(keyFile.toPath(), keyHeader.aesKey());

        // Write the key info (URL, file path, and IV) to the key info file
        String ivHex = bytesToHex(keyHeader.iv());
        String keyInfoContent = "http://example.com/path/to/encryption.key\n" + keyFile.getAbsolutePath() + "\n" + ivHex;
        java.nio.file.Files.write(keyInfoFile.toPath(), keyInfoContent.getBytes());

        return keyInfoFile;
    }

    // TODO: move this to a common helper class
    private String toGuidId(String input) throws NoSuchAlgorithmException {
        // Compute the MD5 hash
        MessageDigest md = MessageDigest.getInstance("MD5");
        byte[] md5Bytes = md.digest(input.getBytes());

        // Convert the first 16 bytes of the MD5 hash to a UUID
        long msb = 0;
        long lsb = 0;
        for (int i = 0; i < 8; i++) {
            msb = (msb << 8) | (md5Bytes[i] & 0xff);
        }
        for (int i = 8; i < 16; i++) {
            lsb = (lsb << 8) | (md5Bytes[i] & 0xff);
        }

        return new UUID(msb, lsb).toString();
    }
}
